multicluster/controllers/multicluster/gateway_controller.go

```diff
@@ -0,0 +1,328 @@
+//+kubebuilder:rbac:groups=multicluster.crd.antrea.io,resources=clusterinfoimports/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=multicluster.crd.antrea.io,resources=clusterinfoimports/finalizers,verbs=update
+
+func (r *GatewayReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	klog.V(2).InfoS("Reconciling Gateway", "gateway", req.NamespacedName)
+	requeue, commonArea, err := r.getCommonArea()
+	if requeue {
+		return ctrl.Result{Requeue: true}, err
+	}
+
+	r.leaderNamespace = commonArea.GetNamespace()
+	err = r.getServiceCIDR(ctx)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	resExportName := r.localClusterID + "-clusterinfo"
+	resExportNamespaced := types.NamespacedName{
+		Name:      resExportName,
+		Namespace: r.leaderNamespace,
+	}
+	resExport := &mcsv1alpha1.ResourceExport{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      resExportName,
+			Namespace: r.leaderNamespace,
+		},
+	}
+
+	createOrUpdate := func(gwIP string) error {
+		existingResExport := &mcsv1alpha1.ResourceExport{}
+		if err := commonArea.Get(ctx, resExportNamespaced, existingResExport); err != nil {
+			if !apierrors.IsNotFound(err) {
+				return err
+			}
+			if err = r.createorUpdateResourceExport(ctx, req, commonArea, existingResExport, gwIP, true); err != nil {
+				return err
+			}
+			return nil
+		}
+		if err = r.createorUpdateResourceExport(ctx, req, commonArea, existingResExport, gwIP, false); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	gwn := &mcsv1alpha1.Gateway{}
+	if err := r.Client.Get(ctx, req.NamespacedName, gwn); err != nil {
+		if apierrors.IsNotFound(err) {
+			gwNodeInfo, err := r.getLastCreatedGateway()
+			if err != nil {
+				klog.ErrorS(err, "Failed to get Gateways")
+				return ctrl.Result{}, err
+			}
+			if gwNodeInfo == nil {
+				// When the last Gateway is deleted, we will remove the ClusterInfo kind of ResourceExport
+				if err := commonArea.Delete(ctx, resExport, &client.DeleteOptions{}); err != nil {
+					return ctrl.Result{}, client.IgnoreNotFound(err)
+				}
+				return ctrl.Result{}, nil
+			}
+			// When there are still Gateways exist, we should create or update existing ResourceExport
+			// with the latest Gateway in remaining Gateways.
+			if err := createOrUpdate(gwNodeInfo.GatewayIP); err != nil {
+				return ctrl.Result{}, err
+			}
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+	if err := createOrUpdate(gwn.GatewayIP); err != nil {
+		return ctrl.Result{}, err
+	}
+	return ctrl.Result{}, nil
+}
+
+// getCommonArea checks ClusterSet initialization status. When ClusterSet is initialized successfully by member
+// ClusterSet controller, remoteCommonAreaManager will be built and include a connected CommonArea.
+func (r *GatewayReconciler) getCommonArea() (requeue bool, remoteCommonArea commonarea.RemoteCommonArea, err error) {
+	defer r.mutex.RUnlock()
+	r.mutex.RLock()
+	if *r.remoteCommonAreaManager == nil {
```

Comment:
When remoteCommonAreaManager can be nil? I did not see where it is set to nil or changed.

And again, I did not see where you lock the mutex for write either.

Reply:
Hi @jianjuns , the remoteCommonAreaManager is initialized here: https://github.com/antrea-io/antrea/blob/627239ce6b124b2594ee8e1d37d69d2d476d0845/multicluster/cmd/multicluster-controller/member.go#L63-L66,  remoteCommonAreaManager will be nil until member_clusterset_controller initialize it when a ClusterSet is setting.

Reply:
Sure, but when GatewayReconciler.remoteCommonAreaManager, if it is initialized to be `&clusterSetReconciler.RemoteCommonAreaManager`, and never set to nil?

And again, your mutex is never write locked? Then why you need it.

Reply:
When member controller is started, the &clusterSetReconciler.RemoteCommonAreaManager is actually nil because no ClusterSet and ClusterClaim yet. When user creates a new ClusterSet, the creation event will be handled here:
https://github.com/antrea-io/antrea/blob/627239ce6b124b2594ee8e1d37d69d2d476d0845/multicluster/controllers/multicluster/member_clusterset_controller.go#L60-L113, then &clusterSetReconciler.RemoteCommonAreaManager will be written and updated, so there is a lock for the whole ClusterSet reconciling process. it will become to nil again when ClusterSet is deleted or leader cluster is disconnected.

Reply:
How about:  `GatewayReconciler.remoteCommonAreaManager` and  `StaleResourceReconciler.remoteCommonAreaManager` (asked 3 times, but still did not see it is answered)?

Reply:
ah, they are actually initialized in the same way in member.go
https://github.com/antrea-io/antrea/blob/c7ca4a2d7344113942ec2ff15860881d6fe715ad/multicluster/cmd/multicluster-controller/member.go#L64-L96.

Reply:
member_clusterset_controller.go is responsible to initialize `remoteCommonAreaManager` when there is a ClusterSet created, or set it to nil when ClusterSet is deleted or leader cluster is disconnected. Other controllers like gateway controller, stale resource controller will only read it and use common area which is managed by `remoteCommonAreaManager`

Reply:
When `GatewayReconciler.remoteCommonAreaManager` can be nit or written? If not why we need lock? If yes, where and why GatewayReconciler.mutex is never write locked?

The same for StaleResourceReconciler.

Reply:
When the member controller is running but no ClusterSet is initialized, `*.remoteCommonAreaManager` will be nil, or when a ClusterSet CR is deleted, it will become nil again. I suppose I can remove the lock in Gateway and stale resource controller's lock since it doesn't update `remoteCommonAreaManager`, only read and use it.

Reply:
I am still confused about your statement. The code you showed me multiple times are about `MemberClusterSetReconciler.remoteCommonAreaManager`. I am asking about `GatewayReconciler.remoteCommonAreaManager` and `StaleResourceReconciler.remoteCommonAreaManager`.

Let me try ask again: from your code (if I did not miss anything) `GatewayReconciler.remoteCommonAreaManager` and `StaleResourceReconciler.remoteCommonAreaManager` are never changed, once GatewayReconciler and StaleResourceReconciler are constructed (and the mutex in GatewayReconciler / StaleResourceReconciler are never write locked). Why you keep saying `*.remoteCommonAreaManager` can be nil?

Reply:
> I suppose I can remove the lock in Gateway and stale resource controller's lock since it doesn't update remoteCommonAreaManager, only read and use it.

This sentence is very confusing to me. We are talking about the pointer to remoteCommonAreaManager in GatewayReconciler / StaleResourceReconciler, not the remoteCommonAreaManager object instance. I also asked you if remoteCommonAreaManager object itself is mutable or not, when GatewayReconciler / StaleResourceReconciler are accessing it (through the saved pointer), but you did not answer that question either. If the answer is YES, the mutex in GatewayReconciler / StaleResourceReconciler does not help at all, and we need to lock the remoteCommonAreaManager object itself, not its references in other structs!

Reply:
Hi @jianjuns, I didn't clarify it clearly, I didn't mean the pointer can be nil, I mean the  remoteCommonAreaManager object  can be nil when ClusterSet is deleted, the code is here https://github.com/antrea-io/antrea/blob/176e9e0c6449fcc033f153322fd33a1cf263e647/multicluster/controllers/multicluster/member_clusterset_controller.go#L69-L81. and the  remoteCommonAreaManager object itself has been locked here https://github.com/antrea-io/antrea/blob/176e9e0c6449fcc033f153322fd33a1cf263e647/multicluster/controllers/multicluster/member_clusterset_controller.go#L67-L68
The method `func (r *MemberClusterSetReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) ` is the only entrypoint when there is any ClusterSet events. `GatewayReconciler.remoteCommonAreaManager` and `StaleResourceReconciler.remoteCommonAreaManager` are referring to the object `MemberClusterSetReconciler.RemoteCommonAreaManager` when they are constructed: e.g: https://github.com/antrea-io/antrea/blob/176e9e0c6449fcc033f153322fd33a1cf263e647/multicluster/cmd/multicluster-controller/member.go#L72-L75
I think RemoteCommonAreaManager object itself is not mutable when GatewayReconciler / StaleResourceReconciler are accessing it since MemberClusterSetReconciler already add a lock during any ClusterSet updating.

Reply:
> the remoteCommonAreaManager object can be nil 

An object cannot never be nil. Only a pointer can be nil.

If GatewayReconciler.remoteCommonAreaManager and StaleResourceReconciler.remoteCommonAreaManager can never be modified, and RemoteCommonAreaManager is not mutable, please remove the mutex in GatewayReconciler and StaleResourceReconciler.

Reply:
Done, I have removed them.